pcsx2: rework

AndersonTorres opened the pull request:
## Description of changes

Merging two packages into one is not a good idea at all. Let's split it.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

matteo-pacini commented:
@AndersonTorres can we get this merged first, as it's been around for a while? 🙏🏻 

https://github.com/NixOS/nixpkgs/pull/324113

superherointj commented on pkgs/by-name/pc/pcsx2/package.nix:
```diff
@@ -1,39 +1,158 @@
+    repo = "pcsx2";
+    fetchSubmodules = true;
+    rev = "v${finalAttrs.version}";
+    sha256 = "sha256-WiwnP5yoBy8bRLUPuCZ7z4nhIzrY8P29KS5ZjErM/A4=";
```
```suggestion
    hash = "sha256-WiwnP5yoBy8bRLUPuCZ7z4nhIzrY8P29KS5ZjErM/A4=";
```

superherointj commented on pkgs/by-name/pc/pcsx2/package.nix:
```diff
@@ -1,39 +1,158 @@
+    zstd
+  ] ++ cubeb.passthru.backendLibs;
+
+  installPhase = ''
```
installPhase hooks are missing

AndersonTorres replied:
I am not fixing the already bad code, I am just undoing the mess of merging a binary-only MacOS-only repack into a source-based build.

AndersonTorres replied:
Now the code is fixed!

superherointj commented on pkgs/by-name/pc/pcsx2/package.nix:
```diff
@@ -1,39 +1,158 @@
+      --run 'if [[ -z $I_WANT_A_BROKEN_WAYLAND_UI ]]; then export QT_QPA_PLATFORM=xcb; fi'
+  '';
+
+  meta = {
```
> changelog = "https://github.com/PCSX2/pcsx2/releases/tag/v${finalAttrs.version}";
> downloadPage = "https://github.com/PCSX2/pcsx2";

AndersonTorres replied:
I forgot this. Thanks, I will retake this when the other PR is merged.

superherointj commented on pkgs/by-name/pc/pcsx2/package.nix:
```diff
@@ -1,39 +1,158 @@
+    owner = "PCSX2";
+    repo = "pcsx2_patches";
+    rev = "b3a788e16ea12efac006cbbe1ece45b6b9b34326";
+    sha256 = "sha256-Uvpz2Gpj533Sr6wLruubZxssoXefQDey8GHIDKWhW3s=";
```
```suggestion
    hash = "sha256-Uvpz2Gpj533Sr6wLruubZxssoXefQDey8GHIDKWhW3s=";
```

AndersonTorres replied:
Done

AndersonTorres commented:
> @AndersonTorres can we get this merged first, as it's been around for a while? 🙏🏻
> 
> #324113

Oh I can.

afh commented:
Thanks for pinging me on this one, @AndersonTorres. Possibly my understanding of a package differs from what nixpkgs considers a package. In my mind a package is a piece of software that can be installed, regardless of whether from source or from binary.

This PR seems to suggest that nixpkgs differentiates between source builds and binary installs. Can you please point me to documentation that outlines the reasons why such a distinction is made? I'm genuinely curious 🙂

superherointj commented:
> Thanks for pinging me on this one, @AndersonTorres. Possibly my understanding of a package differs from what nixpkgs considers a package. In my mind a package is a piece of software that can be installed, regardless of whether from source or from binary.
> 
> This PR seems to suggest that nixpkgs differentiates between source builds and binary installs. Can you please point me to documentation that outlines the reasons why such a distinction is made? I'm genuinely curious 🙂

I believe, the matter is how packages are referenced by-name, not some conception notion of package.

https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md

Two top-level entries, two directories. Is that the issue?

matteo-pacini commented on pkgs/by-name/pc/pcsx2-bin/package.nix:
```diff
@@ -0,0 +1,58 @@
+    ];
+    mainProgram = "pcsx2-qt";
+    maintainers = with lib.maintainers; [
+      hrdinka
```
@AndersonTorres I think I am the only Darwin maintainer, as @hrdinka  and @GovanifY were on the original Linux derivation when I added Darwin support.

https://github.com/NixOS/nixpkgs/pull/320122/files#diff-fcfb4cc512aebed71f6d491a7b51d67b72d2c7ee4fdfe90dbd97ed6d701742a9L148

AndersonTorres replied:
OK, I will remember this

AndersonTorres replied:
Done.

AndersonTorres commented:
@afh 

> In my mind a package is a piece of software that can be installed, regardless of whether from source or from binary.

Yes, let's pick this definition.

> This PR seems to suggest that nixpkgs differentiates between source builds and binary installs. 

The difference is more pronounced. This is a similar situation from vlc vs vlc-bin:

- bin: https://github.com/NixOS/nixpkgs/pull/291405
- src: https://github.com/NixOS/nixpkgs/pull/209086

Both packages follow completely distinct methods of deployment:

One is binary-only, MacOS-only, can't be customized and merely repacked.
The other is effectively deployed from sources, and albeit being Linux-only, it can possibly be customized.

Further, it creates an unreasonable coupling.
Suppose the next release of PCSX2 uses new libraries.
The precompiled file can be released almost immediately, while the script to build the source one needs to be revised and updated.
It makes no sense to couple both releases to the same script.

AndersonTorres committed:
pcsx2: "remove" macos binary package

Let's "remove" the binary-only, MacOS-only package from this expression, because
it should be maintained by MacOS maintainers.

Besides, no other modification will be made right now, just a merge between files.

AndersonTorres committed:
pcsx2-bin: "init" at 1.7.5919

This is a repack of MacOS-only precompiled package from upstream.

Basically it is a migration of the previous Darwin-only package for pcsx2.

By request, matteo-pacini is set as the sole maintainer of this package.

AndersonTorres committed:
pcsx2: fix expression

- lacks runHooks on installPhase
- remove nested with

AndersonTorres committed:
pcsx2: big rewrite

- Gather all the sources in a same expression under sources.nix
- Doc-comment and rename the patch
- strictDeps is set as true
  - Why extra-cmake-modules does not work in nativeBuildInputs??
- Add myself as maintainer

AndersonTorres commented:
Now we have a "small" problem: Darwin users will need to call `pcsx2-bin` instead of merely `pcsx`.

If this is such a big issue, I can rename pcsx2 to pcsx2-src and green-alias both according to the platform.

afh commented:
Possibly I'm misguided by my own misconceptions of nixpkgs or general preconceptions or even both 😅 

First and foremost as written words can easily misread in their tone:
The mindset in which the following was written was friendly, appreciative, cooperative, and constructive. If it is perceived it otherwise for whatever reasons I kindly ask you, esteemed reader, to re-read this comment in the intended spirit and would appreciate feedback where the wording was misunderstood or ill-chosen 🙂 

I think separating packages by platform or installation method, i.e. from source or pre-built binary is heading into the right direction, yet there are several things that don't feel right to me about this as it is currently.

1. I find that a `-bin` or `-src` suffix makes discovering the package harder than it needs to be.
1. The `-bin` or `-src` suffix in this PR seem to be bound to the platform that the package is being installed on. What if in the future nixpkgs would like to offer a upstream-built binary for linux of pcsx2? Where would that reside in the nixpkgs (repo) tree?
1. As a user of a platform—in this case darwin—it does not matter to me whether the package is going to be installed from source or a pre-built binary and I would find it confusing if the package were named differently, in this case `pcsx2-bin` especially if the reason for this is a technical issue, possibly even a limitation of the current state of nixpkgs.
1. As a maintainer or contributor I'd prefer all related Nix code to reside under a single directory (or as much as possible)

I'm very much in favor of the `pkgs/by-name` direction that nixpkgs is heading in (especially with [NixOS/rfc#146](https://github.com/NixOS/rfcs/pull/146) and #325357 being actively worked on 👏) yet I do see value in a os-specific tree. Would it be possible and improve this PR by `git mv pkgs/by-name/pc/pcsx2-bin/package.nix pkgs/by-name/pc/pcsx2/darwin.nix` and adding `pcsx2 = ../pkgs/by-name/pc/pcsx2/darwin.nix { };` to `pkgs/top-level/darwin-packages.nix`?

Another approach could be to `pcsx2 = pkgs.pcsx2-bin;` in pkgs/top-level/darwin-aliases.nix and somehow inform darwin users who are trying to install `pcsx2` that the package is available on their platforms as `darwin.pcsx2` and/or `pcsx2-bin`.

I don't think I properly understand what you mean by "green-alias", @AndersonTorres, could you provide more context and an example, please?

AndersonTorres commented:
> I think separating packages by platform or installation method, i.e. from source or pre-built binary is heading into the right direction, yet there are several things that don't feel right to me about this as it is currently.
> 
>     1. I find that a `-bin` or `-src` suffix makes discovering the package harder than it needs to be.

The idea is to write something like

```nix
pcsx2 = if stdenv.isDarwin then pcsx2-bin else pcsx2-src;
```

> 
>     2. The `-bin` or `-src` suffix in this PR seem to be bound to the platform that the package is being installed on. 

Not exactly. The `-bin` is bound to the platform by its own nature, but `-src` is bound because we didn't figure out how to  build it for both.

> What if in the future nixpkgs would like to offer a upstream-built binary for linux of pcsx2?

In Nixpkgs we prefer source-based builds. The upstream binary package would be rejected.
Indeed we are only including the MacOS binary package because we didn't figure how to build it from sources yet.

The only exception I can imagine is when the binary package has some closed-source extras that are not included by default on the sources.

> I don't think I properly understand what you mean by "green-alias", @AndersonTorres, could you provide more context and an example, please?

By design, aliases put in `pkgs/top-level/aliases.nix` are rejected by OfBorg, in the following sense:  using `pkgconfig`  in your code is fine, but when uploading it to Nixpkgs, ofBorg selectively ignores `aliases.nix` and do not recognize `pkgconfig` as an alias for `pkg-config`.

A "green alias" would be one put in all-packages.nix.

matteo-pacini reviewed (approved):
LGTM - I'm happy with `pcsx2-bin` for Darwin, as it explicitly tells me it's a binary distribution.

superherointj commented:
Result of `nixpkgs-review pr 325562` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

superherointj commented:
Result of `nixpkgs-review pr 325562` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pcsx2</li>
  </ul>
</details>

superherointj commented:
Result of `nixpkgs-review pr 325562` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
